ioctl/cmd/did/diddoc.go

```diff
@@ -15,22 +25,135 @@ const (
+		ID              string `json:"id"`
+		Type            string `json:"type"`
+		Controller      string `json:"controller"`
+		PublicKeyBase58 string `json:"publicKeyBase58,omitempty"`
 	}
+
+	verificationMethodSet interface{}
+
+	serviceStruct struct {
+		ID              string `json:"id,omitempty"`
+		Type            string `json:"type,omitempty"`
+		ServiceEndpoint string `json:"serviceEndpoint,omitempty"`
+	}
+
 	// Doc is the DID document struct
 	Doc struct {
-		Context        string                 `json:"@context,omitempty"`
-		ID             string                 `json:"id,omitempty"`
-		Authentication []authenticationStruct `json:"authentication,omitempty"`
+		Context            interface{}             `json:"@context,omitempty"`
+		ID                 string                  `json:"id,omitempty"`
+		Controller         string                  `json:"controller,omitempty"`
+		VerificationMethod []verificationMethod    `json:"verificationMethod,omitempty"`
+		Authentication     []verificationMethodSet `json:"authentication,omitempty"`
+		AssertionMethod    []verificationMethodSet `json:"assertionMethod,omitempty"`
+		Service            []serviceStruct         `json:"service,omitempty"`
 	}
 )
 
-func newDIDDoc() *Doc {
-	return &Doc{
-		Context: "https://www.w3.org/ns/did/v1",
+// Owner did document owner
+func (doc *Doc) Owner() common.Address {
+	return common.HexToAddress(doc.ID[7:])
+}
+
+// Bytes did document bytes
+func (doc *Doc) Bytes() ([]byte, error) {
+	return json.MarshalIndent(doc, "", "  ")
+}
+
+// JSON did document json
+func (doc *Doc) JSON() (string, error) {
+	data, err := doc.Bytes()
+	if err != nil {
+		return "", err
+	}
+	return string(data), nil
+}
+
+// Hash did document hash
+func (doc *Doc) Hash() ([32]byte, error) {
+	data, err := doc.Bytes()
+	if err != nil {
+		return [32]byte{}, err
+	}
+	return sha256.Sum256(data), nil
+}
+
+// AddService add service to did document
+func (doc *Doc) AddService(tag, serviceType, endpoint string) {
+	id := doc.ID + "#" + tag
+	if doc.Service == nil {
+		doc.Service = []serviceStruct{{
+			ID:              id,
+			Type:            serviceType,
+			ServiceEndpoint: endpoint,
+		}}
+	} else {
+		found := false
+		for _, service := range doc.Service {
+			if service.ID == id {
+				service.Type = serviceType
+				service.ServiceEndpoint = endpoint
+				found = true
+			}
+		}
+		if !found {
+			doc.Service = append(doc.Service, serviceStruct{
+				ID:              id,
+				Type:            serviceType,
+				ServiceEndpoint: endpoint,
+			})
+		}
+	}
+}
+
+// RemoveService remove service from did document
+func (doc *Doc) RemoveService(tag string) error {
+	id := doc.ID + "#" + tag
+	if doc.Service == nil {
+		return errors.New("service not exists")
+	}
+	serices := make([]serviceStruct, len(doc.Service)-1)
+	for i, service := range doc.Service {
+		if service.ID != id {
+			if i == len(serices) {
+				return errors.New("service not exists")
+			}
+			serices[i] = doc.Service[i]
```

Comment:
need to use another `count` and do `services[count] = doc.Service[i]`
otherwise say `doc.Service[2] == id`, it will skip 2
```
serices[0] = doc.Service[0]
serices[1] = doc.Service[1]
skip 2
serices[3] = doc.Service[3]
```
which is not correct

also i think no need to define `services []`, just use `doc.Service`? in the example above remove `doc.Service[2]`, copy `doc.Service[3:]` to `doc.Service[2:]`